### PR TITLE
Add a trained model

### PR DESCRIPTION
**Pull request by simlmx**
Add a trained model

More generally we add support for trained model.

* Define (and type) all the parts of the training pipeline.
* Add a generic training script.
* Simplify the training pipeline.
* Model configs are python classes in `psp/ml/exp_configs`.
* Split the data in train/valid/test. Currently we test on different
  pvs *and* time ranges, as we have too much data anyway.
* Add `RandomPvXDataPipe` to generate training samples in a random
  fashion. At the moment we also use it for testing.
* Unify the variable names in PvDataSource.
* Add the `PvDataSource.ignore_future` safeguard.
* Add the `RecentHistoryModel`, which use data from the previous days
  (at the prediction to of day) and data from the previous minutes.
  Those are fed to a random forest (code for a pytorch neural net is
  also included but the results are not good with it yet). The
  implementation is a bit hacky but it's mostly a baseline.
* Add dependencies to be able to work on GCP.
* Add `pydocstyle` to check the style of docstrings
  (although we disable many checks).

### CODE REVIEW

**Comment by peterdudfield** [comment]
Would you be able to add the the Readme and quick description of the files and folder. Perhaps one sentence for each file? This will help give a good over view of things - thanks

File: psp/tests/data/data_sources/test_pv_data_source.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import numpy as np
+import pytest
+import xarray as xr
+
+from psp.data.data_sources.pv import NetcdfPvDataSource
+
+
+@pytest.fixture
+def pv_data_source(tmp_path):
```
Comment (peterdudfield) [inline psp/tests/data/data_sources/test_pv_data_source.py:11]: worth moving this to a conftest.py file?

File: psp/scripts/eval_model.py [hunk 2]
```diff
@@ -1,114 +1,113 @@
...
 
-def parse_args():
-    parser = argparse.ArgumentParser(__doc__)
-    parser.add_argument("-p", "--pv-data", required=True, help=".netcdf PV data")
-    parser.add_argument(
-        "-n",
-        "--exp-name",
-        help="name of the experiment (used as the name of the dictionary)",
+@click.command()
+@exp_root_opt
```
Comment (peterdudfield) [inline psp/scripts/eval_model.py:29]: is it worth expanding these, witha. description and type,

File: psp/ml/models/recent_history.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+from psp.ml.typings import Batch, Features, X, Y
+from psp.pv import get_irradiance
+from psp.utils.maths import safe_div
+
+
+def to_midnight(ts: datetime) -> datetime:
+    return ts.replace(hour=0, minute=0, second=0, microsecond=0)
+
+
+def tod_as_minutes(ts: datetime) -> float:
```
Comment (peterdudfield) [inline psp/ml/models/recent_history.py:21]: `tod` = time of day, is worth ust calling it `time` or `time_of_day`. I find some abbreviations less helpful

File: psp/ml/models/recent_history.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        return y
+
+    def _get_tod_stats(
+        self,
+        x: X,
+        yesterday_midnight: datetime,
+        future_interval: Tuple[float, float],
+        data: xr.DataArray,
+    ):
+
```
Comment (peterdudfield) [inline psp/ml/models/recent_history.py:59]: This is probably a general comment, but can we add function descriptions? I think I would define this as necessary for production. Doesnt have to be in this PR, could ust chunk in a issue to do this

File: psp/scripts/eval_model.py (hunk 2 above)
Comment (simlmx) [reply psp/scripts/eval_model.py:29]: They are defined in `_options.py` with `type` and `help`

File: psp/tests/data/data_sources/test_pv_data_source.py (hunk 1 above)
Comment (simlmx) [reply psp/tests/data/data_sources/test_pv_data_source.py:11]: Not yet! When multiple test files can use the same `NetcdfPvDataSource` then yes!

**Pull request merged**